File: sdk/kotlin/src/main/kotlin/com/muninndb/client/MuninnClient.kt
package com.muninndb.client

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.SocketTimeoutException
import java.util.concurrent.TimeUnit
import kotlin.math.pow

class MuninnClient(
    val baseUrl: String = "http://localhost:8476",
    val token: String = "",
    val timeoutMs: Long = 30_000L,
    val maxRetries: Int = 3,
    val defaultVault: String = "default",
    httpClient: OkHttpClient? = null
) {
    internal val gson: Gson = GsonBuilder().create()
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    internal val http: OkHttpClient = httpClient ?: OkHttpClient.Builder()
        .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .writeTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .build()

    // ── Write ────────────────────────────────────────────────────────

    suspend fun write(options: WriteOptions): WriteResponse {
        return request("POST", "/api/engrams", body = options, responseType = WriteResponse::class.java)
    }

    suspend fun writeBatch(vault: String = defaultVault, engrams: List<WriteOptions>): BatchWriteResponse {
        val body = BatchWriteBody(engrams.map { it.copy(vault = it.vault ?: vault) })
        return request("POST", "/api/engrams/batch", body = body, responseType = BatchWriteResponse::class.java)
    }

    // ── Read ─────────────────────────────────────────────────────────

    suspend fun read(id: String, vault: String? = null): Engram {
        return request(
            "GET", "/api/engrams/$id",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = Engram::class.java
        )
    }

    // ── Forget ───────────────────────────────────────────────────────

    suspend fun forget(id: String, vault: String? = null, hard: Boolean = false) {
        val v = vault ?: defaultVault
        val params = mutableMapOf<String, String?>("vault" to v)
        if (hard) params["hard"] = "true"
        request("DELETE", "/api/engrams/$id", params = params, responseType = Unit::class.java)
    }

    // ── Activate ─────────────────────────────────────────────────────

    suspend fun activate(options: ActivateOptions): ActivateResponse {
        val opts = if (options.vault == null) options.copy(vault = defaultVault) else options
        return request("POST", "/api/activate", body = opts, responseType = ActivateResponse::class.java)
    }

    // ── Link ─────────────────────────────────────────────────────────

    suspend fun link(options: LinkOptions) {
        val opts = if (options.vault == null) options.copy(vault = defaultVault) else options
        request("POST", "/api/link", body = opts, responseType = Unit::class.java)
    }

    // ── Traverse ─────────────────────────────────────────────────────

    suspend fun traverse(options: TraverseOptions): TraverseResponse {
        val opts = if (options.vault == null) options.copy(vault = defaultVault) else options
        return request("POST", "/api/traverse", body = opts, responseType = TraverseResponse::class.java)
    }

    // ── Evolve ───────────────────────────────────────────────────────

    suspend fun evolve(id: String, newContent: String, reason: String, vault: String? = null): EvolveResponse {
        val body = EvolveBody(vault = vault ?: defaultVault, newContent = newContent, reason = reason)
        return request("POST", "/api/engrams/$id/evolve", body = body, responseType = EvolveResponse::class.java)
    }

    // ── Consolidate ──────────────────────────────────────────────────

    suspend fun consolidate(ids: List<String>, mergedContent: String, vault: String? = null): ConsolidateResponse {
        val body = ConsolidateOptions(vault = vault ?: defaultVault, ids = ids, mergedContent = mergedContent)
        return request("POST", "/api/consolidate", body = body, responseType = ConsolidateResponse::class.java)
    }

    // ── Decide ───────────────────────────────────────────────────────

    suspend fun decide(options: DecideOptions): DecideResponse {
        val opts = if (options.vault == null) options.copy(vault = defaultVault) else options
        return request("POST", "/api/decide", body = opts, responseType = DecideResponse::class.java)
    }

    // ── Restore ──────────────────────────────────────────────────────

    suspend fun restore(id: String, vault: String? = null): RestoreResponse {
        return request(
            "POST", "/api/engrams/$id/restore",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = RestoreResponse::class.java
        )
    }

    // ── Explain ──────────────────────────────────────────────────────

    suspend fun explain(options: ExplainOptions): ExplainResponse {
        val opts = if (options.vault == null) options.copy(vault = defaultVault) else options
        return request("POST", "/api/explain", body = opts, responseType = ExplainResponse::class.java)
    }

    // ── Set State ────────────────────────────────────────────────────

    suspend fun setState(id: String, state: String, reason: String? = null, vault: String? = null): SetStateResponse {
        val body = SetStateBody(vault = vault ?: defaultVault, state = state, reason = reason)
        return request("PUT", "/api/engrams/$id/state", body = body, responseType = SetStateResponse::class.java)
    }

    // ── List Deleted ─────────────────────────────────────────────────

    suspend fun listDeleted(vault: String? = null, limit: Int? = null): ListDeletedResponse {
        val params = mutableMapOf<String, String?>("vault" to (vault ?: defaultVault))
        if (limit != null) params["limit"] = limit.toString()
        return request("GET", "/api/deleted", params = params, responseType = ListDeletedResponse::class.java)
    }

    // ── Retry Enrich ─────────────────────────────────────────────────

    suspend fun retryEnrich(id: String, vault: String? = null): RetryEnrichResponse {
        return request(
            "POST", "/api/engrams/$id/retry-enrich",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = RetryEnrichResponse::class.java
        )
    }

    // ── Contradictions ───────────────────────────────────────────────

    suspend fun contradictions(vault: String? = null): ContradictionsResponse {
        return request(
            "GET", "/api/contradictions",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = ContradictionsResponse::class.java
        )
    }

    // ── Guide ────────────────────────────────────────────────────────

    suspend fun guide(vault: String? = null): String {
        val resp = request(
            "GET", "/api/guide",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = GuideResponseBody::class.java
        )
        return resp.guide
    }

    // ── Stats ────────────────────────────────────────────────────────

    suspend fun stats(vault: String? = null): StatsResponse {
        val params = mutableMapOf<String, String?>()
        if (vault != null) params["vault"] = vault
        return request("GET", "/api/stats", params = params, responseType = StatsResponse::class.java)
    }

    // ── List Engrams ─────────────────────────────────────────────────

    suspend fun listEngrams(vault: String? = null, limit: Int? = null, offset: Int? = null): ListEngramsResponse {
        val params = mutableMapOf<String, String?>("vault" to (vault ?: defaultVault))
        if (limit != null) params["limit"] = limit.toString()
        if (offset != null) params["offset"] = offset.toString()
        return request("GET", "/api/engrams", params = params, responseType = ListEngramsResponse::class.java)
    }

    // ── Get Links ────────────────────────────────────────────────────

    suspend fun getLinks(id: String, vault: String? = null): List<AssociationItem> {
        val resp = request(
            "GET", "/api/engrams/$id/links",
            params = mapOf("vault" to (vault ?: defaultVault)),
            responseType = GetLinksResponse::class.java
        )
        return resp.links
    }

    // ── List Vaults ──────────────────────────────────────────────────

    suspend fun listVaults(): List<String> {
        val resp = request("GET", "/api/vaults", responseType = VaultsResponse::class.java)
        return resp.vaults
    }

    // ── Session ──────────────────────────────────────────────────────

    suspend fun session(vault: String? = null, since: String? = null, limit: Int? = null, offset: Int? = null): SessionResponse {
        val params = mutableMapOf<String, String?>("vault" to (vault ?: defaultVault))
        if (since != null) params["since"] = since
        if (limit != null) params["limit"] = limit.toString()
        if (offset != null) params["offset"] = offset.toString()
        return request("GET", "/api/session", params = params, responseType = SessionResponse::class.java)
    }

    // ── Subscribe (SSE) ──────────────────────────────────────────────

    fun subscribe(vault: String? = null, pushOnWrite: Boolean = true, threshold: Double? = null): Flow<SseEvent> {
        val urlBuilder = (baseUrl + "/api/subscribe").toHttpUrl().newBuilder()
            .addQueryParameter("vault", vault ?: defaultVault)
            .addQueryParameter("push_on_write", pushOnWrite.toString())
        if (threshold != null) {
            urlBuilder.addQueryParameter("threshold", threshold.toString())
        }
        val request = Request.Builder()
            .url(urlBuilder.build())
            .apply { if (token.isNotEmpty()) addHeader("Authorization", "Bearer $token") }
            .addHeader("Accept", "text/event-stream")
            .build()
        return sseFlow(http, request)
    }

    // ── Health ───────────────────────────────────────────────────────

    suspend fun health(): HealthResponse {
        return request("GET", "/api/health", responseType = HealthResponse::class.java)
    }

    // ── Internal ─────────────────────────────────────────────────────

    private suspend fun <T> request(
        method: String,
        path: String,
        params: Map<String, String?> = emptyMap(),
        body: Any? = null,
        responseType: Class<T>
    ): T = withRetry {
        withContext(Dispatchers.IO) {
            val urlBuilder = (baseUrl + path).toHttpUrl().newBuilder()
            params.forEach { (k, v) -> if (v != null) urlBuilder.addQueryParameter(k, v) }
            val url = urlBuilder.build()

            val requestBody = if (body != null) {
                gson.toJson(body).toRequestBody(jsonMediaType)
            } else null

            val request = Request.Builder()
                .url(url)
                .method(method, requestBody ?: if (method == "POST" || method == "PUT") "".toRequestBody(null) else null)
                .apply { if (token.isNotEmpty()) addHeader("Authorization", "Bearer $token") }
                .addHeader("Accept", "application/json")
                .build()

            val response = http.newCall(request).execute()
            response.use { resp ->
                val responseBody = resp.body?.string() ?: ""
                if (!resp.isSuccessful) {
                    throw mapError(resp.code, responseBody)
                }
                if (responseType == Unit::class.java) {
                    @Suppress("UNCHECKED_CAST")
                    Unit as T
                } else {
                    try {
                        gson.fromJson(responseBody, responseType)
                            ?: throw MuninnException.DecodingFailed("Null response body")
                    } catch (e: JsonSyntaxException) {
                        throw MuninnException.DecodingFailed("Failed to decode response: ${e.message}", e)
                    }
                }
            }
        }
    }

    private suspend fun <T> withRetry(block: suspend () -> T): T {
        var attempt = 0
        while (true) {
            try {
                return block()
            } catch (e: MuninnException.ServerError) {
                if (attempt >= maxRetries) throw e
                delay((500L * 2.0.pow(attempt)).toLong())
                attempt++
            } catch (e: SocketTimeoutException) {
                if (attempt >= maxRetries) throw MuninnException.Timeout()
                delay((500L * 2.0.pow(attempt)).toLong())
                attempt++
            } catch (e: IOException) {
                if (attempt >= maxRetries) throw MuninnException.ConnectionFailed(e)
                delay((500L * 2.0.pow(attempt)).toLong())
                attempt++
            }
        }
    }

    private fun mapError(code: Int, body: String): MuninnException {
        val message = extractErrorMessage(body)
        return when (code) {
            400 -> MuninnException.Validation(message)
            401, 403 -> MuninnException.Unauthorized(message)
            404 -> MuninnException.NotFound(message)
            409 -> MuninnException.Conflict(message)
            429 -> MuninnException.ServerError(code, message)
            in 500..599 -> MuninnException.ServerError(code, message)
            else -> MuninnException.ServerError(code, message)
        }
    }

    private fun extractErrorMessage(body: String): String {
        return try {
            val env = gson.fromJson(body, ErrorEnvelope::class.java)
            env?.error?.message ?: body
        } catch (_: Exception) {
            body
        }
    }
}

File: sdk/swift/Sources/MuninnDB/MuninnClient.swift
import Foundation

public final class MuninnClient: @unchecked Sendable {
    private let baseURL: URL
    private let token: String
    private let timeout: TimeInterval
    private let maxRetries: Int
    private let defaultVault: String
    private let session: URLSession

    public init(
        baseURL: URL = URL(string: "http://localhost:8476")!,
        token: String,
        timeout: TimeInterval = 30,
        maxRetries: Int = 3,
        defaultVault: String = "default",
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.token = token
        self.timeout = timeout
        self.maxRetries = maxRetries
        self.defaultVault = defaultVault
        self.session = session
    }

    // MARK: - Public API

    public func write(_ options: WriteOptions) async throws -> WriteResponse {
        return try await _request("POST", path: "/api/engrams", body: options)
    }

    public func writeBatch(vault: String? = nil, engrams: [WriteOptions]) async throws -> BatchWriteResponse {
        let v = vault ?? defaultVault
        let body = BatchWriteRequest(engrams: engrams)
        return try await _request("POST", path: "/api/engrams/batch", queryItems: [URLQueryItem(name: "vault", value: v)], body: body)
    }

    public func read(_ id: String, vault: String? = nil) async throws -> Engram {
        let v = vault ?? defaultVault
        return try await _request("GET", path: "/api/engrams/\(id)", queryItems: [URLQueryItem(name: "vault", value: v)])
    }

    public func forget(_ id: String, vault: String? = nil, hard: Bool = false) async throws {
        let v = vault ?? defaultVault
        // Server exposes DELETE /api/engrams/{id} for both soft and hard delete.
        // Pass hard=true as a query param; the server currently performs soft delete
        // regardless — hard delete support is tracked server-side.
        var queryItems = [URLQueryItem(name: "vault", value: v)]
        if hard {
            queryItems.append(URLQueryItem(name: "hard", value: "true"))
        }
        try await _requestVoid("DELETE", path: "/api/engrams/\(id)", queryItems: queryItems)
    }

    public func activate(_ options: ActivateOptions) async throws -> ActivateResponse {
        return try await _request("POST", path: "/api/activate", body: options)
    }

    public func link(_ options: LinkOptions) async throws {
        try await _requestVoid("POST", path: "/api/link", body: options)
    }

    public func traverse(_ options: TraverseOptions) async throws -> TraverseResponse {
        return try await _request("POST", path: "/api/traverse", body: options)
    }

    public func evolve(id: String, newContent: String, reason: String, vault: String? = nil) async throws -> EvolveResponse {
        let v = vault ?? defaultVault
        let body = EvolveRequest(newContent: newContent, reason: reason, vault: v)
        return try await _request("POST", path: "/api/engrams/\(id)/evolve", body: body)
    }

    public func consolidate(ids: [String], mergedContent: String, vault: String? = nil) async throws -> ConsolidateResponse {
        let opts = ConsolidateOptions(ids: ids, mergedContent: mergedContent, vault: vault ?? defaultVault)
        return try await _request("POST", path: "/api/consolidate", body: opts)
    }

    public func decide(_ options: DecideOptions) async throws -> DecideResponse {
        return try await _request("POST", path: "/api/decide", body: options)
    }

    public func restore(_ id: String, vault: String? = nil) async throws -> RestoreResponse {
        let v = vault ?? defaultVault
        return try await _request(
            "POST",
            path: "/api/engrams/\(id)/restore",
            queryItems: [URLQueryItem(name: "vault", value: v)]
        )
    }

    public func explain(_ options: ExplainOptions) async throws -> ExplainResponse {
        return try await _request("POST", path: "/api/explain", body: options)
    }

    public func setState(id: String, state: String, reason: String? = nil, vault: String? = nil) async throws -> SetStateResponse {
        let body = SetStateRequest(vault: vault ?? defaultVault, state: state, reason: reason)
        return try await _request("PUT", path: "/api/engrams/\(id)/state", body: body)
    }

    public func listDeleted(vault: String? = nil, limit: Int? = nil) async throws -> ListDeletedResponse {
        let v = vault ?? defaultVault
        var items = [URLQueryItem(name: "vault", value: v)]
        if let limit = limit {
            items.append(URLQueryItem(name: "limit", value: String(limit)))
        }
        return try await _request("GET", path: "/api/deleted", queryItems: items)
    }

    public func retryEnrich(_ id: String, vault: String? = nil) async throws -> RetryEnrichResponse {
        let v = vault ?? defaultVault
        return try await _request(
            "POST",
            path: "/api/engrams/\(id)/retry-enrich",
            queryItems: [URLQueryItem(name: "vault", value: v)]
        )
    }

    public func contradictions(vault: String? = nil) async throws -> ContradictionsResponse {
        let v = vault ?? defaultVault
        return try await _request("GET", path: "/api/contradictions", queryItems: [URLQueryItem(name: "vault", value: v)])
    }

    public func guide(vault: String? = nil) async throws -> String {
        let v = vault ?? defaultVault
        let resp: GuideResponse = try await _request("GET", path: "/api/guide", queryItems: [URLQueryItem(name: "vault", value: v)])
        return resp.guide
    }

    public func stats(vault: String? = nil) async throws -> StatsResponse {
        var items: [URLQueryItem]?
        if let v = vault {
            items = [URLQueryItem(name: "vault", value: v)]
        }
        return try await _request("GET", path: "/api/stats", queryItems: items)
    }

    public func listEngrams(vault: String? = nil, limit: Int? = nil, offset: Int? = nil) async throws -> ListEngramsResponse {
        let v = vault ?? defaultVault
        var items = [URLQueryItem(name: "vault", value: v)]
        if let limit = limit {
            items.append(URLQueryItem(name: "limit", value: String(limit)))
        }
        if let offset = offset {
            items.append(URLQueryItem(name: "offset", value: String(offset)))
        }
        return try await _request("GET", path: "/api/engrams", queryItems: items)
    }

    public func getLinks(_ id: String, vault: String? = nil) async throws -> [AssociationItem] {
        let v = vault ?? defaultVault
        let resp: LinksResponse = try await _request(
            "GET",
            path: "/api/engrams/\(id)/links",
            queryItems: [URLQueryItem(name: "vault", value: v)]
        )
        return resp.links
    }

    public func listVaults() async throws -> [String] {
        let resp: VaultsResponse = try await _request("GET", path: "/api/vaults")
        return resp.vaults
    }

    public func session(vault: String? = nil, since: String? = nil, limit: Int? = nil, offset: Int? = nil) async throws -> SessionResponse {
        let v = vault ?? defaultVault
        var items = [URLQueryItem(name: "vault", value: v)]
        if let since = since {
            items.append(URLQueryItem(name: "since", value: since))
        }
        if let limit = limit {
            items.append(URLQueryItem(name: "limit", value: String(limit)))
        }
        if let offset = offset {
            items.append(URLQueryItem(name: "offset", value: String(offset)))
        }
        return try await _request("GET", path: "/api/session", queryItems: items)
    }

    public func subscribe(vault: String? = nil, pushOnWrite: Bool = true, threshold: Double? = nil) -> AsyncThrowingStream<SseEvent, Error> {
        let v = vault ?? defaultVault
        var items = [
            URLQueryItem(name: "vault", value: v),
            URLQueryItem(name: "push_on_write", value: pushOnWrite ? "true" : "false"),
        ]
        if let threshold = threshold {
            items.append(URLQueryItem(name: "threshold", value: String(threshold)))
        }

        var components = URLComponents(url: baseURL.appendingPathComponent("/api/subscribe"), resolvingAgainstBaseURL: false)!
        components.queryItems = items

        var request = URLRequest(url: components.url!)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 0 // SSE streams should not time out

        return makeSubscribeStream(session: session, request: request)
    }

    public func health() async throws -> HealthResponse {
        return try await _request("GET", path: "/api/health")
    }

    // MARK: - Private Helpers

    private func _request<T: Decodable>(
        _ method: String,
        path: String,
        queryItems: [URLQueryItem]? = nil,
        body: (any Encodable)? = nil
    ) async throws -> T {
        let data = try await _executeRequest(method, path: path, queryItems: queryItems, body: body)
        do {
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
        } catch {
            throw MuninnError.decodingFailed(error)
        }
    }

    private func _requestVoid(
        _ method: String,
        path: String,
        queryItems: [URLQueryItem]? = nil,
        body: (any Encodable)? = nil
    ) async throws {
        _ = try await _executeRequest(method, path: path, queryItems: queryItems, body: body)
    }

    private func _executeRequest(
        _ method: String,
        path: String,
        queryItems: [URLQueryItem]?,
        body: (any Encodable)?
    ) async throws -> Data {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        components.queryItems = queryItems

        guard let url = components.url else {
            throw MuninnError.validation("Invalid URL: \(path)")
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = timeout

        if let body = body {
            let encoder = JSONEncoder()
            request.httpBody = try encoder.encode(body)
        }

        var lastError: Error?

        for attempt in 0..<maxRetries {
            do {
                let (data, response) = try await session.data(for: request)

                guard let httpResponse = response as? HTTPURLResponse else {
                    throw MuninnError.connectionFailed(
                        NSError(domain: "MuninnDB", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
                    )
                }

                let statusCode = httpResponse.statusCode

                if (200..<300).contains(statusCode) {
                    return data
                }

                let errorMsg = _extractError(from: data)

                // Retry on 429 or 5xx
                if statusCode == 429 || statusCode >= 500 {
                    lastError = MuninnError.serverError(statusCode, errorMsg)
                    if attempt < maxRetries - 1 {
                        let delay = Double(1 << attempt) * 0.5 // 0.5s, 1s, 2s
                        try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                        continue
                    }
                }

                // Non-retryable errors
                switch statusCode {
                case 401: throw MuninnError.unauthorized(errorMsg)
                case 404: throw MuninnError.notFound(errorMsg)
                case 409: throw MuninnError.conflict(errorMsg)
                case 429: throw MuninnError.serverError(statusCode, errorMsg) // rate limited — retries exhausted
                case 400..<500: throw MuninnError.validation(errorMsg)
                default: throw MuninnError.serverError(statusCode, errorMsg)
                }

            } catch let error as MuninnError {
                throw error
            } catch let error as URLError where error.code == .timedOut {
                throw MuninnError.timeout
            } catch {
                if attempt < maxRetries - 1 {
                    lastError = error
                    let delay = Double(1 << attempt) * 0.5
                    try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                    continue
                }
                throw MuninnError.connectionFailed(error)
            }
        }

        throw lastError ?? MuninnError.connectionFailed(
            NSError(domain: "MuninnDB", code: -1, userInfo: [NSLocalizedDescriptionKey: "Max retries exceeded"])
        )
    }

    private func _extractError(from data: Data) -> String {
        if let errorBody = try? JSONDecoder().decode(ErrorBody.self, from: data),
           let msg = errorBody.error?.message {
            return msg
        }
        return String(data: data, encoding: .utf8) ?? "Unknown error"
    }
}
